homeassistant/components/mysensors.py

```diff
@@ -0,0 +1,216 @@
+        gateway.debug = config[DOMAIN].get(CONF_DEBUG, False)
+
+        def gw_start(event):
+            """Callback to trigger start of gateway and any persistence."""
+            gateway.start()
+            hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP,
+                                 lambda event: gateway.stop())
+            if persistence:
+                for node_id in gateway.sensors:
+                    gateway.event_callback('persistence', node_id)
+
+        hass.bus.listen_once(EVENT_HOMEASSISTANT_START, gw_start)
+
+        return gateway
+
+    # Setup all ports from config
+    global GATEWAYS
+    GATEWAYS = {}
+    conf_gateways = config[DOMAIN][CONF_GATEWAYS]
+    if isinstance(conf_gateways, dict):
+        conf_gateways = [conf_gateways]
+    persistence = config[DOMAIN].get(CONF_PERSISTENCE, True)
+
+    for index, gway in enumerate(conf_gateways):
+        port = gway[CONF_PORT]
+        persistence_file = gway.get(
+            CONF_PERSISTENCE_FILE,
+            hass.config.path('mysensors{}.pickle'.format(index + 1)))
+        GATEWAYS[port] = setup_gateway(
+            port, persistence, persistence_file, version)
+
+    for (component, discovery_service) in DISCOVERY_COMPONENTS:
+        # Ensure component is loaded
+        if not bootstrap.setup_component(hass, component, config):
+            return False
+        # Fire discovery event
+        hass.bus.fire(EVENT_PLATFORM_DISCOVERED, {
+            ATTR_SERVICE: discovery_service,
+            ATTR_DISCOVERED: {}})
+
+    return True
+
+
+def pf_callback_factory(
+        s_types, v_types, devices, add_devices, entity_class):
+    """Return a new callback for the platform."""
+    def mysensors_callback(gateway, node_id):
+        """Callback for mysensors platform."""
+        if gateway.sensors[node_id].sketch_name is None:
+            _LOGGER.info('No sketch_name: node %s', node_id)
+            return
+        # previously discovered, just update state with latest info
+        if node_id in devices:
+            for entity in devices[node_id]:
+                entity.update_ha_state(True)
+            return
+
+        # First time we see this node, detect sensors
+        for child in gateway.sensors[node_id].children.values():
+            name = '{} {}.{}'.format(
+                gateway.sensors[node_id].sketch_name, node_id, child.id)
+
+            for value_type in child.values.keys():
+                if child.type not in s_types or value_type not in v_types:
+                    continue
+
+                devices[node_id].append(
+                    entity_class(gateway, node_id, child.id, name, value_type))
+        if devices[node_id]:
+            _LOGGER.info('adding new devices: %s', devices[node_id])
+            add_devices(devices[node_id])
+        for entity in devices[node_id]:
+            entity.update_ha_state(True)
+    return mysensors_callback
+
+
+class GatewayWrapper(mysensors.SerialGateway):
```

Comment:
Thanks! I should've tested this a bit more. I'm thinking I might wrap the existing serialGateway object then instead of subclassing it.